cuequivariance_torch/cuequivariance_torch/primitives/segmented_polynomial_indexed_linear.py

```diff
@@ -0,0 +1,403 @@
+                for path in d.paths[pid_start:pid_end]
+            ]
+            for pid_start, pid_end in zip(pids[:-1], pids[1:])
+        ]
+        coefficients = []
+        self.coefficients_indices = []
+        n = 0
+        for pid_start, pid_end in zip(pids[:-1], pids[1:]):
+            coeffs = []
+            for path in d.paths[pid_start:pid_end]:
+                coefficients.append(path.coefficients.item())
+                coeffs.append(n)
+                n += 1
+            self.coefficients_indices.append(coeffs)
+        self.register_buffer("coefficients", torch.tensor(coefficients))
+
+    def __repr__(self):
+        return self.repr
+
+    def forward(
+        self,
+        input1: torch.Tensor,
+        input2: torch.Tensor,
+        out_shape: torch.Size,
+        counts: torch.Tensor,
+    ):
+        shape_list = [input1.shape, input2.shape, out_shape, [1]]
+        Z = shape_list[self.Zind[0]][self.Zind[1]]
+        C = shape_list[self.Cind[0]][self.Cind[1]]
+        outputs = [
+            [
+                apply_linear(
+                    input1[
+                        :, self.slices[0][i[0]].start : self.slices[0][i[0]].stop
+                    ].reshape((-1,) + shape[0]),
+                    input2[
+                        :, self.slices[1][i[1]].start : self.slices[1][i[1]].stop
+                    ].reshape((-1,) + shape[1]),
+                    Z,
+                    C,
+                    self.uind,
+                    self.vind,
+                    self.wind,
+                    self.reshape_inputs,
+                    counts,
+                    self.subscripts,
+                    self.coefficients[coef],
+                )
+                for i, shape, coef in zip(ii, shapes, coefs)
+            ]
+            for ii, shapes, coefs in zip(
+                self.indices, self.segment_shapes, self.coefficients_indices
+            )
+        ]
+        return torch.cat(
+            [o[0].reshape(out_shape[0], -1) for o in outputs], dim=-1
+        )  # TODO: More general case?
+
+
+def apply_linear(
+    input1: torch.Tensor,
+    input2: torch.Tensor,
+    Z: int,
+    C: int,
+    uind: List[int],
+    vind: List[int],
+    wind: List[int],
+    reshape_inputs: List[int],
+    counts: torch.Tensor,
+    subscripts: Tuple[str],
+    coefficients: torch.Tensor,
+):
+    from cuequivariance_ops_torch import indexed_linear
+
+    shape_list = [input1.shape, input2.shape, [], [1]]
+    u = shape_list[uind[0]][uind[1]]
+    v = shape_list[vind[0]][vind[1]]
+    w = shape_list[wind[0]][wind[1]]
+    outsh = None
+    if reshape_inputs[0]:
+        outsh = input1.shape[:-1]
+        input1 = input1.reshape(-1, input1.shape[-1])
+    if reshape_inputs[1]:
+        outsh = input2.shape[:-1]
+        input2 = input2.reshape(-1, input2.shape[-1])
+    output = indexed_linear(
+        input1, input2, counts * w, u, v, C, Z * w, subscripts, coefficients
+    )
```

Comment:
I need it add a `math_dtype` support (or we need to add a math_dtype=None that has a new meaning => follow the dtype of the input/output)